### PR TITLE
Support running tests when opening a non-base test file

### PR DESCRIPTION
**Pull request by lunakoly**
Otherwise opening .fir.kt instead
of .kt will show no runners
which is annoying since most of
the time git diff contains changes
in K2 test data, so you have to
do extra clicks to come to the base
K1 file.

### CODE REVIEW

**Comment by lunakoly** [comment]
Fixes this:
image

**Comment by KvanTTT** [comment]
It looks like I fixed this problem a week ago: https://github.com/demiurg906/test-data-helper-plugin/pull/32 (https://github.com/demiurg906/test-data-helper-plugin/issues/31).

Also, have you fixed a similar problem with `.kt` and `.args` files pair?

File: src/org/jetbrains/kotlin/test/helper/actions/GeneratedTestComboBoxAction.kt [hunk 1]
```diff
@@ -264,7 +274,7 @@ class GeneratedTestComboBoxAction(val baseEditor: TextEditor) : ComboBoxAction()
                 }
                 val methodPath = File(methodPathComponents.reversed().joinToString("/"))
                     .canonicalFile.absolutePath
-                if (methodPath == truePath) {
+                if (methodPath.asPathWithoutAllExtensions == truePath.asPathWithoutAllExtensions) {
```
Comment (KvanTTT) [inline src/org/jetbrains/kotlin/test/helper/actions/GeneratedTestComboBoxAction.kt:277]: It's better to extract `truePath.asPathWithoutAllExtensions` out of loop since it performs some calculations.

File: src/org/jetbrains/kotlin/test/helper/actions/GeneratedTestComboBoxAction.kt [hunk 2]
```diff
@@ -128,9 +129,18 @@ class GeneratedTestComboBoxAction(val baseEditor: TextEditor) : ComboBoxAction()
                 .submit(AppExecutorUtil.getAppExecutorService())
         }
 
+        private val VirtualFile.nameWithoutAllExtensions get() = name.split('.').first()
+
+        private val String.asPathWithoutAllExtensions: String
+            get() {
+                val pathParts = split(File.separator).toMutableList()
+                pathParts[pathParts.lastIndex] = pathParts[pathParts.lastIndex].split('.').first()
+                return pathParts.joinToString(File.separator)
```
Comment (KvanTTT) [inline src/org/jetbrains/kotlin/test/helper/actions/GeneratedTestComboBoxAction.kt:138]: This looks a bit more optimal without unnecessary allocations:

```suggestion
        val lastFileSeparatorIndex = lastIndexOf(File.separator).let { if (it == -1) 0 else it }
        val extensionDotIndex = indexOf('.', lastFileSeparatorIndex).let { if (it == -1) length else it }
        return substring(0, extensionDotIndex)
```

File: src/org/jetbrains/kotlin/test/helper/actions/GeneratedTestComboBoxAction.kt [hunk 3]
```diff
@@ -128,9 +129,18 @@ class GeneratedTestComboBoxAction(val baseEditor: TextEditor) : ComboBoxAction()
                 .submit(AppExecutorUtil.getAppExecutorService())
         }
 
+        private val VirtualFile.nameWithoutAllExtensions get() = name.split('.').first()
```
Comment (KvanTTT) [inline src/org/jetbrains/kotlin/test/helper/actions/GeneratedTestComboBoxAction.kt:132]: It looks like this extension can be replaced with `asPathWithoutAllExtensions`.

**Review by KvanTTT — CHANGES_REQUESTED** [review]
In my [PR](https://github.com/demiurg906/test-data-helper-plugin/pull/32) extensions are treated strictly (`.fir.kt`, `.args`), but maybe your approach is more general.

**Comment by lunakoly** [comment]
Oh, you also have some changes on the matter :)

File: src/org/jetbrains/kotlin/test/helper/actions/GeneratedTestComboBoxAction.kt (hunk 1 above)
Comment (lunakoly) [reply src/org/jetbrains/kotlin/test/helper/actions/GeneratedTestComboBoxAction.kt:277]: True. Though it unlikely to be a performance issue

File: src/org/jetbrains/kotlin/test/helper/actions/GeneratedTestComboBoxAction.kt (hunk 2 above)
Comment (lunakoly) [reply src/org/jetbrains/kotlin/test/helper/actions/GeneratedTestComboBoxAction.kt:138]: And possibly if we were to do `indexOfLast()` from back to front, then we could also avoid the second `.let{}`, because the result would be subtractable from `lastIndex`: in case of -1 it's `lastIndex - (-1) == length`, and then it just decreases down to give us the proper index. Thought that may be a bit unreadable already :)

Comment (KvanTTT) [reply src/org/jetbrains/kotlin/test/helper/actions/GeneratedTestComboBoxAction.kt:138]: Ok, it's reasonable.

Comment (lunakoly) [reply src/org/jetbrains/kotlin/test/helper/actions/GeneratedTestComboBoxAction.kt:138]: Or maybe not. WDYT? The example below does still perform an allocation, but we can always implement a manual `reversedIndexOf()` if we want

```
val String.asPathWithoutAllExtensions: String
	get() {
    	val indexOfSeparator = indexOf(File.separator)//.also(::println)

        val indexOfFirstExtensionDot = reversed()
            .substring(0, lastIndex - indexOfSeparator)
            .lastIndexOf('.')
            //.also(::println)

        return substring(0, lastIndex - indexOfFirstExtensionDot)
	}
```

Comment (KvanTTT) [reply src/org/jetbrains/kotlin/test/helper/actions/GeneratedTestComboBoxAction.kt:138]: Your last example works incorrectly because it doesn't consider dots in a folder name. For instance, something like this: `F:\tests.2\dontReportWarningAboutChangingExecutionOrderForVararg.fir.kt`.

Also, it looks too compilcated to me (call `reverse` looks confused for such a simple task).

Maybe it makes sense to write some tests on this extension to get rid of potential mistakes.

Comment (lunakoly) [reply src/org/jetbrains/kotlin/test/helper/actions/GeneratedTestComboBoxAction.kt:138]: > Your last example works incorrectly because it doesn't consider dots in a folder name. For instance, something like this: `F:\tests.2\dontReportWarningAboutChangingExecutionOrderForVararg.fir.kt`.

True, my typo. There should have been `val indexOfSeparator = lastIndexOf(File.separator)`, not `indexOf(File.separator)`

File: src/org/jetbrains/kotlin/test/helper/actions/GeneratedTestComboBoxAction.kt (hunk 1 above)
Comment (KvanTTT) [reply src/org/jetbrains/kotlin/test/helper/actions/GeneratedTestComboBoxAction.kt:277]: Sure it's not a performance issue, but just a rule of good manners.

File: src/org/jetbrains/kotlin/test/helper/actions/GeneratedTestComboBoxAction.kt (hunk 2 above)
Comment (lunakoly) [reply src/org/jetbrains/kotlin/test/helper/actions/GeneratedTestComboBoxAction.kt:138]: I understand why this code may seem unintuitive, but you can think of it this way: in order to avoid having `-1`-related corner cases we must make sure that higher values of `indexOf()`-like searching functions correspond directly to shifting indices in one direction.

- The first thing we do is we search for the last file separator, and the more separators we have, the bigger should be the resulting index.
- The second thing we do is we search for the first `'.'` after the last separator which is the same as looking for the last `'.'` before the first separator, and the more extensions we have, the further we should move back from the end of the string.

So, if we for some reason do some `indexOf()` such that its higher values correspond to higher movement along some direction, but `-1` corresponds to the maximum movement along it, then it's a sign that we can look from the other side to possibly get rid of some corner-cased `ifs`.

But I do realize that this intuition may be uncommon

Comment (lunakoly) [reply src/org/jetbrains/kotlin/test/helper/actions/GeneratedTestComboBoxAction.kt:138]: The argument about string allocations is valid, but, again, this is not performance-critical code, and even if it were, there is a straghtforward way to implement `reversedIndexOf(char, startIndex)` which would do exactly the same as `indexOf()`, but return the index from the end rather than from the start.

Comment (lunakoly) [reply src/org/jetbrains/kotlin/test/helper/actions/GeneratedTestComboBoxAction.kt:138]: A pity that we don't have such a function in our stdlib, though

File: src/org/jetbrains/kotlin/test/helper/actions/GeneratedTestComboBoxAction.kt (hunk 3 above)
Comment (lunakoly) [reply src/org/jetbrains/kotlin/test/helper/actions/GeneratedTestComboBoxAction.kt:132]: On the second thought, since you consider lack of attention to performance details good manners, then why do you suggest removing this function and using a slower one instead?

Comment (lunakoly) [reply src/org/jetbrains/kotlin/test/helper/actions/GeneratedTestComboBoxAction.kt:132]: In theory, the other function needs to deal with file separators on top of just removing the extensions, which could be more work

File: src/org/jetbrains/kotlin/test/helper/actions/GeneratedTestComboBoxAction.kt (hunk 2 above)
Comment (KvanTTT) [reply src/org/jetbrains/kotlin/test/helper/actions/GeneratedTestComboBoxAction.kt:138]: Ok, let's wait for a final resolution from @demiurg906 or @cypressious 😊

File: src/org/jetbrains/kotlin/test/helper/actions/GeneratedTestComboBoxAction.kt (hunk 3 above)
Comment (KvanTTT) [reply src/org/jetbrains/kotlin/test/helper/actions/GeneratedTestComboBoxAction.kt:132]: > On the second thought, since you consider lack of attention to performance details good manners, then why do you suggest removing this function and using a potentially slower one instead?

In this case I choose a bit overhead over function duplication.

File: src/org/jetbrains/kotlin/test/helper/actions/GeneratedTestComboBoxAction.kt (hunk 1 above)
Comment (lunakoly) [reply src/org/jetbrains/kotlin/test/helper/actions/GeneratedTestComboBoxAction.kt]: Applied your suggestion in [c9c6ef2](https://github.com/demiurg906/test-data-helper-plugin/pull/34/commits/c9c6ef2145f539bd394df60c7bf3db07a78c85b1)

File: src/org/jetbrains/kotlin/test/helper/actions/GeneratedTestComboBoxAction.kt (hunk 2 above)
Comment (lunakoly) [reply src/org/jetbrains/kotlin/test/helper/actions/GeneratedTestComboBoxAction.kt]: I have just realized that `reversedIndexOf()` itself can be implemented via `lastIndex - indexOfOrLength()` and the latter function reads way easier (in fact, it's you have inlined in your suggestion). Left it as a separate function, because at this point I believe it deserves it. See [c9c6ef2](https://github.com/demiurg906/test-data-helper-plugin/pull/34/commits/c9c6ef2145f539bd394df60c7bf3db07a78c85b1)

File: src/org/jetbrains/kotlin/test/helper/actions/GeneratedTestComboBoxAction.kt (hunk 3 above)
Comment (lunakoly) [reply src/org/jetbrains/kotlin/test/helper/actions/GeneratedTestComboBoxAction.kt]: See: [c9c6ef2](https://github.com/demiurg906/test-data-helper-plugin/pull/34/commits/c9c6ef2145f539bd394df60c7bf3db07a78c85b1)

**Review by KvanTTT — APPROVED** [review]
